Test/paulmck/rcu.hpp

```diff
@@ -3,45 +3,26 @@
 #include <cstddef>
 #include <memory>
 #include <mutex>
+#include <tuple>
 #include <utility>
 
 // Derived-type approach.  All RCU-protected data structures using this
 // approach must derive from std::rcu_obj_base, which in turn derives
 // from std::rcu_head.  No idea what happens in case of multiple inheritance.
 
 namespace std {
-    template<typename T, typename D = default_delete<T>, bool E = is_empty<D>::value>
-    class rcu_obj_base: private rcu_head {
-        D deleter;
+    template<typename T, typename D = default_delete<T>>
+    class rcu_obj_base: private rcu_head, private std::tuple<D> {
```

Comment:
I didn't find anything else in the standard library, but it looks like boost::compressed_pair will do EBO: https://wandbox.org/permlink/xbT0JaR5UNy4ZOzQ

I suppose it could be used like this:

    namespace detail {
        struct rcu_obj_base_empty_type {};
    }

    template<typename T, typename D = default_delete<T>>
    class rcu_obj_base:
        private rcu_head,
        private boost::compressed_pair<D, detail::rcu_obj_base_empty_type> {
        // ...

Otherwise, I suppose that we will need to roll our own.